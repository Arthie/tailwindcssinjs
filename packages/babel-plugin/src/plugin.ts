import Babel, { PluginObj, PluginPass } from "@babel/core";
import {
  requireTailwindConfig,
  resolveTailwindConfigPath,
  TailwindConfig,
} from "@tailwindcssinjs/tailwindcss-data";
import { tailwindcssinjs } from "./tailwindcssinjs";

/**
 * tries to get tailwind config and stores config in state
 * if it fails it stores default config in state
 * @param state
 * @param config
 */
function getTailwindConfig(config: string) {
  try {
    const resolvedConfigPath = resolveTailwindConfigPath(config);
    return requireTailwindConfig(resolvedConfigPath);
  } catch (err) {
    return requireTailwindConfig(); //returns default config
  }
}

//todo: check if file exists if not error

export interface PluginConfig {
  output?: string;
  config?: string;
}

// https://stackoverflow.com/a/32749533/971592
class PluginError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "@tailwindcssinjsPluginError";
    if (typeof Error.captureStackTrace === "function") {
      Error.captureStackTrace(this, this.constructor);
    } else if (!this.stack) {
      this.stack = new Error(message).stack;
    }
  }
}

interface PluginState {
  outputPath: string;
  configPath: string;
  importDefaultSpecifier?: string;
  config: TailwindConfig;
  tailwind: any;
}

export default function (
  { types: t }: typeof Babel,
  config: PluginConfig
): PluginObj<PluginPass> {
  let pluginState: PluginState;
  return {
    name: "tailwindcssinjs",
    pre(state) {
      // console.log("PRE", state)
      // console.log("gen config", pluginState)
      const outputPath = config.output;
      const configPath = config.config ?? "./tailwind.config.js";
      if (!outputPath)
        throw new PluginError("Add output to @tailwindcssinjs plugin options");
      const tailwindConfig = getTailwindConfig(configPath);
      pluginState = {
        outputPath,
        configPath,
        config: tailwindConfig,
        tailwind: tailwindcssinjs(tailwindConfig, outputPath),
      };
    },
    visitor: {
      ImportDeclaration(path, state) {
        // console.log("Import", state)
        const source = path.get("source");
        if (source.node.value === "@tailwindcssinjs/css") {
          const specifiers = path.get("specifiers");
          const importDefaultSpecifier = specifiers.find(
            (specifier) => specifier.type === "ImportDefaultSpecifier"
          ) as Babel.NodePath<Babel.types.ImportDefaultSpecifier> | undefined;
          if (importDefaultSpecifier) {
            pluginState.importDefaultSpecifier = importDefaultSpecifier.get(
              "local"
            ).node.name;
            importDefaultSpecifier.remove();
          }
          if (path.get("specifiers").length === 0) {
            path.remove();
          }
        }
      },
      TaggedTemplateExpression(path, state) {
        const tag = path.get("tag");
        if (
          tag.node.type === "Identifier" &&
          pluginState.importDefaultSpecifier &&
          tag.node.name === pluginState.importDefaultSpecifier
        ) {
          const quasi = path.get("quasi");
          const templateElements = quasi
            .get("quasis")
            .map((item) => item.node.value.raw);

          const expressions = quasi
            .get("expressions")
            .map((item) => item.evaluate().value);
          const twClasses = [];
          while (templateElements.length || expressions.length) {
            const twClassString = templateElements.shift();
            const twClassObject = expressions.shift();
            if (twClassString) {
              twClasses.push(twClassString);
            }
            if (twClassObject) {
              twClasses.push(twClassObject);
            }
          }
          // console.log("TAG", state, twClasses)
          path.replaceWith(
            t.stringLiteral(pluginState.tailwind.setClasses(twClasses))
          );
        }
      },
    },
    post(state) {
      // console.log("POST", state)
      if (pluginState.outputPath) {
        pluginState.tailwind.writeCSS();
      }
    },
  };
}
