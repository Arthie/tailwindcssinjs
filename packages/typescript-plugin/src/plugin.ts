import {
  decorateWithTemplateLanguageService,
  Logger,
  TemplateContext,
  TemplateLanguageService,
} from "typescript-template-language-service-decorator";
import * as ts from "typescript/lib/tsserverlibrary";
import * as vscode from "vscode-languageserver-types";
import { TwObject } from "@tailwindcssinjs/transformers";
import { TailwindConfig } from "@tailwindcssinjs/tailwindcss-data/lib/tailwindcssConfig";

import tailwindcssinjs, { requireTailwindConfig } from "./tailwindcssinjs";
import { createLogger } from "./logger";
import {
  TailwindContext,
  getTailwindContextFromPosition,
  getTailwindContextFromTemplateContext,
} from "./tailwindContext";
import {
  translateCompletionItemsToCompletionEntryDetails,
  translateCompletionEntry,
  translateHoverItemsToQuickInfo,
} from "./translate";
import postcss from "postcss";

let tailwindConfig: TailwindConfig | undefined;
let tailwindData:
  | {
      twObjectMap: Map<string, TwObject>;
      screens: string[];
      variants: string[];
      generateCssFromText: (text: string) => postcss.Root[];
    }
  | undefined;

let watcher: ts.FileWatcher | undefined;

function updateTailwindcssinjs(logger: Logger, configPath?: string) {
  try {
    logger.log(`CONFIG PATH:${configPath}`);
    if (configPath) {
      if (watcher) {
        watcher.close();
      }
      watcher = ts.sys.watchFile?.(
        configPath,
        (fileName: string, eventKind: ts.FileWatcherEventKind) => {
          logger.log(`Watchfile config update: ${configPath}`);
          updateTailwindcssinjs(logger, fileName);
        }
      );
    }
    tailwindConfig = requireTailwindConfig(logger, configPath);
    tailwindData = tailwindcssinjs(tailwindConfig);
  } catch (err) {
    logger.log(`ERROR: ${err}`);
  }
}

function tailwindcssinjsLanguageService(
  info: ts.server.PluginCreateInfo,
  logger: Logger
): TemplateLanguageService {
  try {
    updateTailwindcssinjs(logger, info.config?.config);

    return {
      getSyntacticDiagnostics(context: TemplateContext): ts.Diagnostic[] {
        if (!tailwindData) {
          return [];
        }
        const maxNumberOfProblems = 100;
        let problems = 0;
        const diagnostics: ts.Diagnostic[] = [];

        //TODO improve this regex
        const NESTED_VARIANT_REGEXP = /(\[([^\[\]]){0,}\[[^\]]{0,}\][^\]]{0,}\])|(\[([^\[\]]){0,}\[)|(\]([^\[\]]){0,}\])/;

        const nestedMath = context.text.match(NESTED_VARIANT_REGEXP);

        if (nestedMath?.length) {
          problems++;
          let diagnostic: ts.Diagnostic = {
            file: context.node.getSourceFile(),
            code: 1,
            category: ts.DiagnosticCategory.Error,
            start: nestedMath.index,
            length: nestedMath[0].length,
            messageText: `Nested variants arrays are not allowed.`,
            source: "tailwindcssinjs",
          };
          diagnostics.push(diagnostic);
        }

        const tailwindContext = getTailwindContextFromTemplateContext(
          context,
          tailwindConfig?.separator ?? ":"
        );

        for (const templateContextClass of tailwindContext) {
          if (problems > maxNumberOfProblems) break;

          if (templateContextClass.type === "array") {
            for (const twClass of templateContextClass.classes) {
              const twObject = tailwindData.twObjectMap.get(twClass.text);
              if (!twObject) {
                diagnostics.push({
                  file: context.node.getSourceFile(),
                  code: 2,
                  category: ts.DiagnosticCategory.Error,
                  start: twClass.index,
                  length: twClass.text.length,
                  messageText: `${twClass.text} is not a tailwind class.`,
                  source: "tailwindcssinjs",
                });
                problems++;
              }
            }
          }

          if (templateContextClass.type === "variant") {
            const twObject = tailwindData.twObjectMap.get(
              templateContextClass.class.text
            );
            if (!twObject) {
              diagnostics.push({
                file: context.node.getSourceFile(),
                code: 2,
                category: ts.DiagnosticCategory.Error,
                start: templateContextClass.class.index,
                length: templateContextClass.class.text.length,
                messageText: `${templateContextClass.class.text} is not a tailwind class.`,
                source: "tailwindcssinjs",
              });
              problems++;
            }
          }

          if (templateContextClass.type === "class") {
            if (!tailwindData.twObjectMap.has(templateContextClass.text)) {
              diagnostics.push({
                file: context.node.getSourceFile(),
                code: 5,
                category: ts.DiagnosticCategory.Error,
                start: templateContextClass.index,
                length: templateContextClass.text.length,
                messageText: `${templateContextClass.text} is not a tailwind class.`,
                source: "tailwindcssinjs",
              });
              problems++;
            }
          }

          if (
            templateContextClass.type === "array" ||
            templateContextClass.type === "variant"
          ) {
            //first:second:...other
            const [first, second, ...other] = templateContextClass.variant;

            const getVariantType = (variant?: string) => {
              if (variant) {
                if (tailwindData?.screens.includes(variant)) {
                  return "screen";
                }
                if (tailwindData?.variants.includes(variant)) {
                  return "variants";
                }
              }
            };

            const firstType = getVariantType(first);
            const secondType = getVariantType(second);

            // if (other.length) {
            //   diagnostics.push({
            //     file: context.node.getSourceFile(),
            //     code: 6,
            //     category: ts.DiagnosticCategory.Error,
            //     start: templateContextClass.index,
            //     length: templateContextClass.variant.join(
            //       tailwindConfig?.separator ?? ":"
            //     ).length,
            //     messageText: `${templateContextClass.variant.join(
            //       tailwindConfig?.separator ?? ":"
            //     )} is not a valid tailwind variant. The selector has more than 2 variants: ${other}.`,
            //     source: "tailwindcssinjs",
            //   });
            //   problems++;
            //   continue;
            // }

            // if (secondType && firstType === "variants") {
            //   diagnostics.push({
            //     file: context.node.getSourceFile(),
            //     code: 7,
            //     category: ts.DiagnosticCategory.Error,
            //     start: templateContextClass.index,
            //     length: templateContextClass.variant.join(
            //       tailwindConfig?.separator ?? ":"
            //     ).length,
            //     messageText: `${templateContextClass.variant.join(
            //       tailwindConfig?.separator ?? ":"
            //     )} is not a valid tailwind variant. "${first}" should be a screen variant`,
            //     source: "tailwindcssinjs",
            //   });
            //   problems++;
            //   continue;
            // }

            if (!firstType && !secondType) {
              diagnostics.push({
                file: context.node.getSourceFile(),
                code: 8,
                category: ts.DiagnosticCategory.Error,
                start: templateContextClass.index,
                length: templateContextClass.variant.join(
                  tailwindConfig?.separator ?? ":"
                ).length,
                messageText: `${templateContextClass.variant.join(
                  tailwindConfig?.separator ?? ":"
                )} is not a tailwind variant.`,
                source: "tailwindcssinjs",
              });
              problems++;
              continue;
            }

            if (firstType === "screen" && secondType === "screen") {
              diagnostics.push({
                file: context.node.getSourceFile(),
                code: 9,
                category: ts.DiagnosticCategory.Error,
                start: templateContextClass.index,
                length: templateContextClass.variant.join(
                  tailwindConfig?.separator ?? ":"
                ).length,
                messageText: `${templateContextClass.variant.join(
                  tailwindConfig?.separator ?? ":"
                )} is not a valid tailwind variant. "${second}" should be a variant, not a screen variant`,
                source: "tailwindcssinjs",
              });
              problems++;
              continue;
            }
          }
        }

        return diagnostics;
      },
      getCompletionsAtPosition(
        context: TemplateContext,
        position: ts.LineAndCharacter
      ): ts.CompletionInfo {
        if (!tailwindData) {
          return {
            isGlobalCompletion: false,
            isMemberCompletion: false,
            isNewIdentifierLocation: false,
            entries: [],
          };
        }

        const tailwindContexts = getTailwindContextFromTemplateContext(
          context,
          tailwindConfig?.separator ?? ":"
        );

        const twContext = getTailwindContextFromPosition(
          tailwindContexts,
          position
        );

        //todo add check for is tailwind class
        const entries: ts.CompletionEntry[] = [];
        for (const key of tailwindData.twObjectMap.keys()) {
          const entry = translateCompletionEntry({
            label: key,
            kind: vscode.CompletionItemKind.Text,
          });

          if (twContext) {
            if (twContext.type === "array") {
              const twContextClass = getTailwindContextFromPosition(
                twContext.classes,
                position
              );
              if (twContextClass) {
                entry.replacementSpan = {
                  start: twContextClass.index,
                  length: twContextClass.text.length,
                };
              }
            } else {
              const endsWithSeparator = twContext.text.endsWith(
                tailwindConfig?.separator ?? ":"
              );
              if (!endsWithSeparator) {
                entry.replacementSpan = {
                  start: twContext.index,
                  length: twContext.text.length,
                };
              }
            }
          }
          entries.push(entry);
        }

        for (const variant of [
          ...tailwindData.screens,
          ...tailwindData.variants,
        ]) {
          const variantEntry = translateCompletionEntry({
            label: variant,
            kind: vscode.CompletionItemKind.Function,
          });
          entries.push(variantEntry);
        }

        return {
          isGlobalCompletion: false,
          isMemberCompletion: false,
          isNewIdentifierLocation: false,
          entries: entries,
        };
      },
      getCompletionEntryDetails(
        context: TemplateContext,
        position: ts.LineAndCharacter,
        name: string
      ): ts.CompletionEntryDetails {
        if (!tailwindData) {
          return {
            displayParts: [],
            kind: ts.ScriptElementKind.unknown,
            kindModifiers: "",
            name,
          };
        }

        const twObject = tailwindData.twObjectMap.get(name);
        if (twObject) {
          const root = postcss.root().append(...twObject.nodes);
          return translateCompletionItemsToCompletionEntryDetails({
            label: name,
            kind: vscode.CompletionItemKind.Text,
            documentation: {
              kind: vscode.MarkupKind.Markdown,
              value: [
                "```css",
                root.toString().replace(/    /g, "  "),
                "```",
              ].join("\n"),
            },
          });
        }
        return {
          displayParts: [],
          kind: ts.ScriptElementKind.unknown,
          kindModifiers: "",
          name,
        };
      },
      getQuickInfoAtPosition(
        context: TemplateContext,
        position: ts.LineAndCharacter
      ): ts.QuickInfo | undefined {
        if (!tailwindData) {
          return;
        }

        const tailwindContexts = getTailwindContextFromTemplateContext(
          context,
          tailwindConfig?.separator ?? ":"
        );

        const twContext = getTailwindContextFromPosition(
          tailwindContexts,
          position
        );

        const getQuickInfo = (
          twContext: TailwindContext,
          twObjectRoots: postcss.Root[]
        ) => {
          return translateHoverItemsToQuickInfo({
            start: twContext.index,
            label: twContext.text,
            kind: vscode.CompletionItemKind.Text,
            documentation: {
              kind: vscode.MarkupKind.Markdown,
              value: [
                "```css",
                ...twObjectRoots.map((twObjectRoot: postcss.Root) => [
                  twObjectRoot.toString().replace(/    /g, "  "),
                ]),
                "```",
              ].join("\n"),
            },
          });
        };

        if (twContext) {
          const twObjectRoots = tailwindData.generateCssFromText(
            twContext.text
          );
          return getQuickInfo(twContext, twObjectRoots);
        }
      },
    };
  } catch (err) {
    logger.log(`ERROR: ${err}`);
    throw err;
  }
}

interface PluginConfig {
  config: string;
}

export = (mod: { typescript: typeof ts }): ts.server.PluginModule => {
  let logger: Logger;
  return {
    create(info: ts.server.PluginCreateInfo): ts.LanguageService {
      logger = createLogger(info);
      logger.log("TailwindcssinjsTsPlugin is starting.");
      return decorateWithTemplateLanguageService(
        mod.typescript,
        info.languageService,
        info.project,
        tailwindcssinjsLanguageService(info, logger),
        { tags: ["tw"] },
        { logger }
      );
    },
    onConfigurationChanged(pluginConfig: PluginConfig) {
      if (logger) {
        logger.log(
          `Config has changed ${typeof pluginConfig} ${JSON.stringify(
            pluginConfig
          )}`
        );
      }
      updateTailwindcssinjs(logger, pluginConfig.config);
    },
  };
};
