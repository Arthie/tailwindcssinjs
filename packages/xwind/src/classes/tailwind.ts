import core, {
  createTwClassDictionary,
  ResolvedTailwindConfig,
} from "@xwind/core";

import initTwClassesUtils, {
  TwClasses,
  TwParsedClass,
} from "@xwind/class-utilities";

function tailwind(
  tailwindConfig: ResolvedTailwindConfig,
  includeBase: boolean
) {
  const {
    resolvedConfig,
    utilitiesRoot,
    componentsRoot,
    baseRoot,
    screens,
    variants,
    generateTwClassSubstituteRoot,
  } = core(tailwindConfig);

  const twClassDictionary = createTwClassDictionary(
    componentsRoot,
    utilitiesRoot
  );

  const baseCSSroot = baseRoot.clone();
  baseCSSroot.walkAtRules("layer", (atRule) => {
    atRule.replaceWith(atRule.nodes);
  });

  const compare = (
    { class: firstClass, variants: firstVariants }: TwParsedClass,
    { class: secondClass, variants: secondVariants }: TwParsedClass
  ) => {
    const firstIsScreen = screens.indexOf(firstVariants[0]);
    const secondIsScreen = screens.indexOf(secondVariants[0]);
    if (firstIsScreen !== -1 || secondIsScreen !== -1) {
      if (firstIsScreen < secondIsScreen) {
        return -1;
      }
      if (firstIsScreen > secondIsScreen) {
        return 1;
      }
      return 0;
    }

    if (firstClass < secondClass) {
      return -1;
    }

    if (firstClass > secondClass) {
      return 1;
    }
    return 0;
  };

  const twClassesUtils = initTwClassesUtils(resolvedConfig.separator, [
    ...screens,
    ...variants,
  ]);

  const generatedTwClassesCSS: { [key: string]: string } = {};

  return (twClasses: TwClasses) => {
    const sortedTwClasses = twClassesUtils.parser(twClasses).sort(compare);
    const combinedRoot: string[] = [];
    for (const twClass of sortedTwClasses) {
      const [generatedTwClass] = twClassesUtils.generator(twClass);

      let generatedTwClassCSS = generatedTwClassesCSS[generatedTwClass];

      if (!generatedTwClassCSS) {
        generatedTwClassCSS = generateTwClassSubstituteRoot(
          twClassDictionary,
          twClass
        ).toString();
        generatedTwClassesCSS[generatedTwClass] = generatedTwClassCSS;
      }
      combinedRoot.push(generatedTwClassCSS);
    }

    if (!includeBase) {
      return ["/* Generated by @xwind/babel-plugin */", ...combinedRoot].join(
        "\n"
      );
    }
    return [
      "/* Generated by @xwind/babel-plugin */",
      baseCSSroot.toString(),
      ...combinedRoot,
    ].join("\n");
  };
}

export default tailwind;
