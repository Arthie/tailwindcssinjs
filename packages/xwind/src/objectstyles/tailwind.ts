import {
  createTwClassDictionary,
  tailwindData,
  ResolvedTailwindConfig,
  transformTwRootToObjectStyle,
  mergeObjectStyles,
  ObjectStyle,
} from "@xwind/core";

import initTwClassesUtils, { TwClasses } from "@xwind/class-utilities";

function tailwind(tailwindConfig: ResolvedTailwindConfig) {
  const {
    resolvedConfig,
    utilitiesRoot,
    componentsRoot,
    screens,
    variants,
    generateTwClassSubstituteRoot,
  } = tailwindData(tailwindConfig);

  const twClassDictionary = createTwClassDictionary(
    componentsRoot,
    utilitiesRoot
  );

  const twClassesUtils = initTwClassesUtils(resolvedConfig.separator, [
    ...screens,
    ...variants,
  ]);

  return (twClasses: TwClasses) => {
    const parsedTwClasses = twClassesUtils.parser(twClasses);
    const composedTwClasses = twClassesUtils.composer(twClasses);

    const objectStyles: ObjectStyle[] = [];
    for (const parsedTwClass of parsedTwClasses) {
      const twRoot = generateTwClassSubstituteRoot(
        twClassDictionary,
        parsedTwClass
      );
      objectStyles.push(transformTwRootToObjectStyle(parsedTwClass[0], twRoot));
    }

    let objectStyle = mergeObjectStyles(objectStyles);

    if (resolvedConfig.xwind?.objectstyles?.plugins) {
      for (const plugin of resolvedConfig.xwind.objectstyles.plugins) {
        objectStyle = plugin(objectStyle, composedTwClasses, resolvedConfig);
      }
    }

    return objectStyle;
  };
}

export default tailwind;
