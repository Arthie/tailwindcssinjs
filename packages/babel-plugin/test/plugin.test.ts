import pluginTester from "babel-plugin-tester";

import plugin from "../lib/plugin";

pluginTester({
  plugin,
  pluginName: "tailwindcssinjs",
  pluginOptions: {
    output: "./test.css",
  },
  tests: [
    {
      title: "remove import",
      code: 'import tw from "tailwindcssinjs"',
      output: "",
    },
    {
      title: "other import",
      code: 'import { cx } from "tailwindcssinjs";',
      output: 'import { cx } from "tailwindcssinjs";',
    },
    {
      title: "other import2",
      code: `
        import tailwind from "tailwindcssinjs"
        const test = tailwind\`bg-red-200\`
      `,
      output: 'const test = "bg-red-200";',
    },
    {
      title: "test 1",
      code: `
        import tw from "tailwindcssinjs"
        const test = tw\`bg-red-200\`
      `,
      output: 'const test = "bg-red-200";',
    },
    {
      title: "test 2",
      code: `
        import tw from "tailwindcssinjs"
        const test = tw\`bg-red-200 text-sm\`
        const test2 = tw\`bg-red-200 text-sm\`
      `,
      snapshot: true,
    },
  ],
});
