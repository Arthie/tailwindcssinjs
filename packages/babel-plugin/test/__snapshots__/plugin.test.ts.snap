// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tailwindcssinjs test 2: test 2 1`] = `

import tw from "tailwindcssinjs"
const test = tw\`bg-red-200 text-sm\`
const test2 = tw\`bg-red-200 text-sm\`

      ↓ ↓ ↓ ↓ ↓ ↓

const test = "bg-red-200 text-sm";
const test2 = "bg-red-200 text-sm";


`;
